src/coreclr/jit/valuenum.cpp

```diff
@@ -5618,6 +5634,56 @@ bool ValueNumStore::IsVNInt32Constant(ValueNum vn)
     return TypeOfVN(vn) == TYP_INT;
 }
 
+bool ValueNumStore::IsVNNeverNegative(ValueNum vn)
+{
+    if (IsVNConstant(vn))
+    {
+        var_types vnTy = TypeOfVN(vn);
+        if (varTypeIsIntegral(vnTy))
```

Comment:
Probably faster if you check the type first thing and bail out quickly if it's not an integer.

Reply:
Ah you are checking it at the call site... so then why check again here?

Reply:
I could put an assert here for the type and we can always rely on the call-site to be responsible for what type it is.